internal/datanode/buffer.go

```diff
@@ -172,26 +170,28 @@ func (m *DeltaBufferManager) ShouldFlushSegments() []UniqueID {
 	var (
 		poppedSegmentIDs []UniqueID
 		poppedItems      []*Item
+		bufferSize       int64
 	)
-	for {
+	m.heapGuard.Lock()
+	for memUsage-bufferSize >= Params.DataNodeCfg.FlushDeleteBufferBytes {
 		segItem := m.popHeapItem()
 		poppedItems = append(poppedItems, segItem)
 		poppedSegmentIDs = append(poppedSegmentIDs, segItem.segmentID)
-		memUsage -= segItem.memorySize
-		if memUsage < Params.DataNodeCfg.FlushDeleteBufferBytes {
-			break
-		}
+		bufferSize += segItem.memorySize
 	}
 
 	//here we push all selected segment back into the heap
 	//in order to keep the heap semantically correct
-	m.heapGuard.Lock()
 	for _, segMem := range poppedItems {
```

Comment:
There is something peculiar here, pop everything out and then push it back in.
Subsequently, it may be advisable to modify the data structure to avoid such inefficient operations.